spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraPartitionedRDD.scala

```diff
@@ -22,15 +32,42 @@ class CassandraPartitionedRDD[T](prev: RDD[T])(implicit ct: ClassTag[T]) extends
   override def getPartitions: Array[Partition] = {
     partitioner match {
       case Some(rp: ReplicaPartitioner) => prev.partitions.map(partition => rp.getEndpointPartition(partition))
-      case _ => throw new IllegalArgumentException("CassandraPartitionedRDD hasn't been partitioned by ReplicaPartitioner. This should be impossible")
+      case _ => throw new IllegalArgumentException("CassandraPartitionedRDD hasn't been " +
+        "partitioned by ReplicaPartitioner. Unable to do any work with data locality.")
     }
   }
 
+  /**
+   * This method currently uses thrift to determine the local endpoint and rpc endpoints
+   * from whatever endpoint the partition belongs to. This gives us a better chance of matching
+   * the bound interface of the Spark Executor. In addition we will add the HostName and
+   * HostAddress that we get for each of these endpoints to cover all of the logical choices.
+   */
   override def getPreferredLocations(split: Partition): Seq[String] = {
+    val connector = CassandraConnector(prev.sparkContext.getConf)
+
+    val rpcToLocalAddress = connector.withCassandraClientDo { client =>
+      val ring = client.describe_local_ring(keyspace)
+      for {
+        tr <- ring
+        rpcIps = tr.rpc_endpoints.map(InetAddress.getByName)
+        localIps = tr.endpoints.map(InetAddress.getByName)
+        (rpc, local) <- (rpcIps).zip(localIps)
+      } yield (rpc, local)
+    }.toMap
+    val localToRpcAddress = rpcToLocalAddress.map { case (k, v) => (v -> k) }
+
     split match {
       case epp: ReplicaPartition =>
-        epp.endpoints.map(_.getHostAddress).toSeq
-      case other: Partition => throw new IllegalArgumentException("CassandraPartitionedRDD doesn't have Endpointed Partitions. This should be impossible.")
+        epp.endpoints.flatMap { origInet =>
```

Comment:
Because it seems this is not necessary to even execute unless we receive a type ReplicaPartition, I would move this here